crates/hyperdrive-math/src/lp.rs

```diff
@@ -4,9 +4,144 @@ use ethers::types::{I256, U256};
 use fixed_point::FixedPoint;
 use fixed_point_macros::{fixed, int256};
 
-use crate::{State, YieldSpace};
+use crate::{calculate_effective_share_reserves, State, YieldSpace};
 
 impl State {
+    // Calculates the lp_shares for a given contribution when adding liquidity.
+    pub fn calculate_add_liquidity(
+        &self,
+        current_block_timestamp: U256,
+        contribution: FixedPoint,
+        min_apr: FixedPoint,
+        max_apr: FixedPoint,
+        as_base: bool,
+    ) -> Result<FixedPoint, &str> {
+        // Ensure that the contribution is greater than or equal to the minimum
+        // transaction amount.
+        if contribution < self.minimum_transaction_amount() {
+            return Err("MinimumTransactionAmount");
```

Comment:
we're using Eyre for our error messages, not just Err. No idea what the difference is, but this is the established pattern:

```rs
if self.y() < dy {
    return Err(eyre!(
        "calculate_shares_in_given_bonds_out_up_safe: y = {} < {} = dy",
        self.y(),
        dy,
    ));
}
```

Reply:
nice call, this cleans up the return signature of the function.